content/ghc-9.12-cabal-3.14/index.md

```diff
@@ -0,0 +1,126 @@
++++
+title = "GHC 9.12 & cabal 3.14 releases"
+date = 2024-12-22
+[taxonomies]
+authors = ["Hécate"]
+categories = ["Cabal", "GHC"]
+tags = ["Release"]
++++
+
+To conclude the year 2024, the GHC and Cabal teams are happy to announce the release of GHC 9.12 and cabal 3.14.
+
+<!-- more -->
+
+
+Here are some highlights:
+
+## GHC 9.12
+
+### Language Extensions
+
+This release brings its lot of new and exciting extensions:
+* [Multiline string literals][MultilineStrings] to spread a piece of text over several lines;
+* [Or-Patterns][OrPatterns], allowing you to match on several patterns in a single `case` branch:
+
+```haskell
+data Sweet = Cupcake | Liquorice | Cookie | Raisins
+
+-- Without Or-Patterns
+tasty Cupcake = True
+tasty Cookie = True
+tasty _ = False
+
+-- With Or-Patterns
+tasty (Cupcake; Cookie) = True
+tasty (Liquorice; Raisins) = False
+```
+
+* [NamedDefaults][NamedDefaults] allows you to create type defaulting declarations for literals other than for the `Num` class. This mechanism is used to make sure that in the expression `print (6 + 7)`, the multiplication gets a concrete type like `Integer`, `Int` or `Double`, instead of the vague  `Num a => a`.  
+You can now apply this mechanism to other typeclasses, like:
+
+```haskell
+default IsString (Text, String)
+```
+
+Which means that in a module with [`OverloadedStrings`][OverloadedStrings], string literals `"like this"` will default to [`Text`][Text] instead of a vague polymorphic type.
+
+### Base Library
+
+* Improvements to backtraces and exceptions. Lots of them.
+
+#### ⚠️ Deprecations
+
+* The deprecation cycle of [GHC.Pack][GHC.Pack] has come to an end. Goodbye!
+* [GHC.Desugar][GHC.Desugar] will be removed in GHC 9.14.
+
+### Code Generation
+
+* Experimental support for the [RISC-V Platform](https://gitlab.haskell.org/ghc/ghc/-/issues/16783);
+* SIMD! In the x86: Most floating-point operations and some integer vector operations are supported 128-bit vectors **without LLVM**. [Get in touch](https://matrix.to/#/#ghc:matrix.org) to help with this effort
+* You can try the new experimental [`-fobject-determinism`][-fobject-determinism] flag to enable deterministic object code generation.
+
+#### ⚠️ Deprecations
+
+* Support 32-bit Windows & macOS/iOS has been dropped;
+* As a result, friendship with the `stdcall` calling convention for FFI is ended. Now `ccall` is your best friend.
```

Comment:
```suggestion
* As a result, friendship with the `stdcall` calling convention for FFI has ended. Now `ccall` is your best friend.
```

Reply:
Hm, actually don't we want to recommend `CApiFFI` and `capi`, come to think of it?

Reply:
It's not like you can replace `stdcall` with `ccall` and keep compiling on 32-bit Windows: the support for Win32 ended, thus `stdcall` serves no purpose any longer.

Reply:
True, but that also rules out the recommendation for `ccall`. In any case, I think the deprecation period for 32-bit Windows ending makes that moot: if you want to continue using it you must use an older ghc, which will still support `stdcall`.

Reply:
TBH I never understood why Microsoft chose to use the Pascal calling convention when nobody to speak of was using a Pascal compiler to write programs for Windows. Yes, strictly speaking it has some advantages, but they're outweighed by the added complexity of needing to switch calling conventions for system APIs.

Reply:
@bgamari @andreaspk what's the recommended phrasing here? Should we steer users away from `ccall` and directly to `capi`, or is `ccall` a good intermediate step when they migrate away from `stdcall`?

Reply:
I'll settle for `ccall`.